File: kfcore/src/main/java/com/lq/ren/base/ui/obs/Observable.java
package com.lq.ren.base.ui.obs;

import java.util.ArrayList;
import java.util.List;

public class Observable<T> {

    List<Observer<T>> observers = new ArrayList<>();

    boolean changed = false;

    public void addObserver(Observer<T> observer) {
        if (observer == null) {
            throw new NullPointerException("observer == null");
        }
        synchronized (this) {
            if (!observers.contains(observer))
                observers.add(observer);
        }
    }

    protected void clearChanged() {
        changed = false;
    }

    public synchronized void deleteObserver(Observer<T> observer) {
        observers.remove(observer);
    }

    public synchronized void deleteObservers() {
        observers.clear();
    }

    public boolean hasChanged() {
        return changed;
    }

    protected void setChanged() {
        changed = true;
    }

    public void notifyObservers() {
        notifyObserver(null);
    }

    @SuppressWarnings("unchecked")
    public void notifyObserver(T data) {
        int size = 0;
        Observer<T>[] arrays = null;
        synchronized (this) {
            if (hasChanged()) {
                clearChanged();
                size = observers.size();
                arrays = new Observer[size];
                observers.toArray(arrays);
            }
        }
        if (arrays != null) {
            for (Observer<T> observer : arrays) {
                observer.update(this, data);
            }
        }
    }

}

File: kfcore/src/main/java/com/lq/ren/base/ui/obs/Observable.kt
package com.lq.ren.base.ui.obs

import java.util.ArrayList

open class Observable<T> {

    internal var observers: MutableList<Observer<T>> = ArrayList()

    internal var changed = false

    open fun addObserver(observer: Observer<T>?) {
        if (observer == null) {
            throw NullPointerException("observer == null")
        }
        synchronized(this) {
            if (!observers.contains(observer))
                observers.add(observer)
        }
    }

    protected fun clearChanged() {
        changed = false
    }

    @Synchronized
    fun deleteObserver(observer: Observer<T>) {
        observers.remove(observer)
    }

    @Synchronized
    fun deleteObservers() {
        observers.clear()
    }

    fun hasChanged(): Boolean {
        return changed
    }

    protected fun setChanged() {
        changed = true
    }

    open fun notifyObservers() {
        notifyObserver(null)
    }

    open fun notifyObserver(data: T?) {
        var size = 0
        var arrays: Array<Observer<T>>? = null
        synchronized(this) {
            if (hasChanged()) {
                clearChanged()
                size = observers.size
                arrays = arrayOfNulls<Observer<*>>(size)
                observers.toTypedArray<Observer<T>>()
            }
        }
        if (arrays != null) {
            for (observer in arrays!!) {
                observer.update(this, data)
            }
        }
    }

}
